File: src/dialogs.cpp
#include "dialogs.h"
#include <QFileInfo>
#include <QVBoxLayout>
#include <QHBoxLayout>
#include <QFormLayout>
#include <QPushButton>
#include <QDialogButtonBox>
#include <QFileDialog>
#include <QGroupBox>
#include <QRadioButton>
#include <QButtonGroup>
#include <QStackedWidget>
#include <QLabel>
#include <QColorDialog>

// ── MergeDialog ───────────────────────────────────────────────────────────────

MergeDialog::MergeDialog(QWidget* parent) : QDialog(parent) {
    setWindowTitle("Merge PDFs");
    setMinimumSize(420, 320);

    auto* root   = new QVBoxLayout(this);
    m_list = new QListWidget(this);
    m_list->setDragDropMode(QAbstractItemView::InternalMove);
    m_list->setDefaultDropAction(Qt::MoveAction);

    auto* btnRow    = new QHBoxLayout;
    auto* btnAdd    = new QPushButton("Add PDF…");
    auto* btnRemove = new QPushButton("Remove");
    auto* btnUp     = new QPushButton("▲ Up");
    auto* btnDown   = new QPushButton("▼ Down");
    btnRow->addWidget(btnAdd);
    btnRow->addWidget(btnRemove);
    btnRow->addWidget(btnUp);
    btnRow->addWidget(btnDown);
    btnRow->addStretch();

    connect(btnAdd,    &QPushButton::clicked, this, &MergeDialog::onAdd);
    connect(btnRemove, &QPushButton::clicked, this, &MergeDialog::onRemove);
    connect(btnUp,     &QPushButton::clicked, this, &MergeDialog::onMoveUp);
    connect(btnDown,   &QPushButton::clicked, this, &MergeDialog::onMoveDown);

    auto* btns = new QDialogButtonBox(QDialogButtonBox::Ok | QDialogButtonBox::Cancel);
    connect(btns, &QDialogButtonBox::accepted, this, &QDialog::accept);
    connect(btns, &QDialogButtonBox::rejected, this, &QDialog::reject);

    root->addLayout(btnRow);
    root->addWidget(m_list, 1);
    root->addWidget(btns);
}

QStringList MergeDialog::selectedFiles() const {
    QStringList out;
    for (int i = 0; i < m_list->count(); ++i)
        out << m_list->item(i)->data(Qt::UserRole).toString();
    return out;
}

void MergeDialog::onAdd() {
    QStringList files = QFileDialog::getOpenFileNames(
        this, "Select PDFs", {}, "PDF files (*.pdf)");
    for (const auto& f : files) {
        auto* it = new QListWidgetItem(QFileInfo(f).fileName());
        it->setData(Qt::UserRole, f);
        m_list->addItem(it);
    }
}

void MergeDialog::onRemove() {
    qDeleteAll(m_list->selectedItems());
}

void MergeDialog::onMoveUp() {
    int row = m_list->currentRow();
    if (row <= 0) return;
    auto* it = m_list->takeItem(row);
    m_list->insertItem(row - 1, it);
    m_list->setCurrentRow(row - 1);
}

void MergeDialog::onMoveDown() {
    int row = m_list->currentRow();
    if (row < 0 || row >= m_list->count() - 1) return;
    auto* it = m_list->takeItem(row);
    m_list->insertItem(row + 1, it);
    m_list->setCurrentRow(row + 1);
}

// ── SplitDialog ───────────────────────────────────────────────────────────────

SplitDialog::SplitDialog(int totalPages, QWidget* parent)
    : QDialog(parent), m_totalPages(totalPages)
{
    setWindowTitle("Split PDF");
    setMinimumWidth(380);

    auto* root   = new QVBoxLayout(this);
    auto* grp    = new QGroupBox("Split mode", this);
    auto* grpLay = new QVBoxLayout(grp);

    auto* rByRanges  = new QRadioButton("By page ranges");
    auto* rEveryN    = new QRadioButton("Every N pages");
    auto* rEachPage  = new QRadioButton("Each page as separate file");
    rByRanges->setChecked(true);

    auto* modeGroup = new QButtonGroup(this);
    modeGroup->addButton(rByRanges,  0);
    modeGroup->addButton(rEveryN,    1);
    modeGroup->addButton(rEachPage,  2);

    // Ranges widget
    m_rangesWidget = new QWidget;
    auto* rfl   = new QFormLayout(m_rangesWidget);
    m_ranges = new QLineEdit;
    m_ranges->setPlaceholderText(QString("e.g. 1-3,4-%1").arg(totalPages));
    rfl->addRow("Ranges:", m_ranges);
    rfl->addRow(new QLabel(QString("(Total pages: %1)").arg(totalPages)));

    // EveryN widget
    m_everyNWidget = new QWidget;
    auto* enl = new QFormLayout(m_everyNWidget);
    m_n = new QSpinBox;
    m_n->setRange(1, qMax(1, totalPages - 1));
    m_n->setValue(1);
    enl->addRow("Pages per chunk:", m_n);

    auto* stack = new QStackedWidget(this);
    stack->addWidget(m_rangesWidget);
    stack->addWidget(m_everyNWidget);
    stack->addWidget(new QLabel("Each page saved as Page_1.pdf, Page_2.pdf…"));

    connect(modeGroup, &QButtonGroup::idClicked, stack, &QStackedWidget::setCurrentIndex);

    grpLay->addWidget(rByRanges);
    grpLay->addWidget(rEveryN);
    grpLay->addWidget(rEachPage);
    grpLay->addWidget(stack);

    auto* btns = new QDialogButtonBox(QDialogButtonBox::Ok | QDialogButtonBox::Cancel);
    connect(btns, &QDialogButtonBox::accepted, this, &QDialog::accept);
    connect(btns, &QDialogButtonBox::rejected, this, &QDialog::reject);

    root->addWidget(grp);
    root->addWidget(btns);

    // keep for mode() query
    m_modeEachPage = new QCheckBox; m_modeEachPage->setVisible(false); m_modeEachPage->setParent(this);
    m_modeEveryN   = new QCheckBox; m_modeEveryN->setVisible(false);   m_modeEveryN->setParent(this);
    connect(rEveryN,   &QRadioButton::toggled, m_modeEveryN,   &QCheckBox::setChecked);
    connect(rEachPage, &QRadioButton::toggled, m_modeEachPage, &QCheckBox::setChecked);
}

SplitDialog::Mode SplitDialog::mode() const {
    if (m_modeEachPage->isChecked()) return EachPage;
    if (m_modeEveryN->isChecked())   return EveryN;
    return ByRanges;
}

QString SplitDialog::rangeText() const { return m_ranges->text(); }
int     SplitDialog::everyN()    const { return m_n->value(); }

// ── AddTextDialog ─────────────────────────────────────────────────────────────

AddTextDialog::AddTextDialog(const QString& detectedPdfFont,
                             double detectedSize,
                             QWidget* parent)
    : QDialog(parent)
    , m_color(Qt::black)
{
    setWindowTitle("Add Text");
    setMinimumWidth(400);

    auto* root = new QVBoxLayout(this);
    auto* form = new QFormLayout;

    // Text input
    m_textEdit = new QPlainTextEdit(this);
    m_textEdit->setFixedHeight(90);
    m_textEdit->setPlaceholderText("Type text here…");
    form->addRow("Text:", m_textEdit);

    // Font family
    m_familyCombo = new QComboBox(this);
    m_familyCombo->addItem("Helvetica (Sans-serif)",  "Helvetica");
    m_familyCombo->addItem("Times (Serif)",           "Times");
    m_familyCombo->addItem("Courier (Monospace)",     "Courier");

    // Bold / Italic
    m_boldCheck   = new QCheckBox("Bold",   this);
    m_italicCheck = new QCheckBox("Italic", this);

    auto* styleRow = new QHBoxLayout;
    styleRow->addWidget(m_familyCombo);
    styleRow->addWidget(m_boldCheck);
    styleRow->addWidget(m_italicCheck);
    styleRow->addStretch();
    form->addRow("Font:", styleRow);

    // Size
    m_sizeBox = new QDoubleSpinBox(this);
    m_sizeBox->setRange(4.0, 288.0);
    m_sizeBox->setSingleStep(0.5);
    m_sizeBox->setSuffix(" pt");
    m_sizeBox->setValue(detectedSize > 0 ? detectedSize : 12.0);
    form->addRow("Size:", m_sizeBox);

    // Color
    m_colorBtn = new QPushButton(this);
    m_colorBtn->setFixedWidth(80);
    connect(m_colorBtn, &QPushButton::clicked, this, &AddTextDialog::pickColor);
    updateColorButton();
    form->addRow("Color:", m_colorBtn);

    root->addLayout(form);

    auto* btns = new QDialogButtonBox(QDialogButtonBox::Ok | QDialogButtonBox::Cancel);
    connect(btns, &QDialogButtonBox::accepted, this, &QDialog::accept);
    connect(btns, &QDialogButtonBox::rejected, this, &QDialog::reject);
    root->addWidget(btns);

    // Pre-populate from detected font name
    // Parse "Helvetica-BoldOblique", "Times-Bold", "Courier-Oblique", etc.
    const QString dn = detectedPdfFont.toLower();
    if (dn.contains("times") || dn.contains("roman"))
        m_familyCombo->setCurrentIndex(1);
    else if (dn.contains("courier") || dn.contains("mono"))
        m_familyCombo->setCurrentIndex(2);
    else
        m_familyCombo->setCurrentIndex(0);

    m_boldCheck->setChecked(dn.contains("bold"));
    m_italicCheck->setChecked(dn.contains("italic") || dn.contains("oblique"));
}

QString AddTextDialog::text() const {
    return m_textEdit->toPlainText();
}

QString AddTextDialog::pdfFontName() const {
    const QString family = m_familyCombo->currentData().toString();
    const bool bold   = m_boldCheck->isChecked();
    const bool italic = m_italicCheck->isChecked();

    if (family == "Times") {
        if (bold && italic) return "Times-BoldItalic";
        if (bold)           return "Times-Bold";
        if (italic)         return "Times-Italic";
        return "Times-Roman";
    }
    if (family == "Courier") {
        if (bold && italic) return "Courier-BoldOblique";
        if (bold)           return "Courier-Bold";
        if (italic)         return "Courier-Oblique";
        return "Courier";
    }
    // Helvetica (default)
    if (bold && italic) return "Helvetica-BoldOblique";
    if (bold)           return "Helvetica-Bold";
    if (italic)         return "Helvetica-Oblique";
    return "Helvetica";
}

double AddTextDialog::fontSize() const { return m_sizeBox->value(); }
QColor AddTextDialog::color()    const { return m_color; }

void AddTextDialog::pickColor() {
    QColor c = QColorDialog::getColor(m_color, this, "Text Color");
    if (c.isValid()) { m_color = c; updateColorButton(); }
}

void AddTextDialog::updateColorButton() {
    m_colorBtn->setStyleSheet(
        QString("background-color: %1; border: 1px solid #666;").arg(m_color.name()));
    m_colorBtn->setText(m_color.name().toUpper());
}

// ── CropDialog ────────────────────────────────────────────────────────────────

static QDoubleSpinBox* makeMarginSpin() {
    auto* s = new QDoubleSpinBox;
    s->setRange(0, 500);
    s->setSingleStep(1.0);
    s->setSuffix(" pt");
    return s;
}

CropDialog::CropDialog(QWidget* parent) : QDialog(parent) {
    setWindowTitle("Crop Page");
    setMinimumWidth(300);

    auto* root   = new QVBoxLayout(this);
    auto* form   = new QFormLayout;

    m_top    = makeMarginSpin();
    m_bottom = makeMarginSpin();
    m_left   = makeMarginSpin();
    m_right  = makeMarginSpin();
    form->addRow("Top margin (pt):",    m_top);
    form->addRow("Bottom margin (pt):", m_bottom);
    form->addRow("Left margin (pt):",   m_left);
    form->addRow("Right margin (pt):",  m_right);

    m_allPages = new QCheckBox("Apply to all pages", this);
    form->addRow(m_allPages);

    auto* btns = new QDialogButtonBox(QDialogButtonBox::Ok | QDialogButtonBox::Cancel);
    connect(btns, &QDialogButtonBox::accepted, this, &QDialog::accept);
    connect(btns, &QDialogButtonBox::rejected, this, &QDialog::reject);

    root->addLayout(form);
    root->addWidget(btns);
}

double CropDialog::top()      const { return m_top->value();    }
double CropDialog::bottom()   const { return m_bottom->value(); }
double CropDialog::left()     const { return m_left->value();   }
double CropDialog::right()    const { return m_right->value();  }
bool   CropDialog::allPages() const { return m_allPages->isChecked(); }

File: pdfmaestro/dialogs.py
"""
Qt dialogs for PDF operations: Merge, Split, Crop.
Each dialog collects user input and exposes it as properties.
Actual pikepdf work is done in operations.py.
"""

from __future__ import annotations
from PySide6.QtWidgets import (
    QDialog, QVBoxLayout, QHBoxLayout, QFormLayout,
    QLabel, QPushButton, QListWidget, QListWidgetItem,
    QRadioButton, QButtonGroup, QSpinBox, QDoubleSpinBox,
    QLineEdit, QFileDialog, QDialogButtonBox, QGroupBox,
    QComboBox, QAbstractItemView, QMessageBox,
)
from PySide6.QtCore import Qt


# ── Merge dialog ──────────────────────────────────────────────────────────────

class MergeDialog(QDialog):
    """
    Lets the user pick additional PDFs and choose where to insert them.
    The current document is always the base; extra files are appended/prepended.
    """

    def __init__(self, current_path: str, page_count: int, parent=None):
        super().__init__(parent)
        self.setWindowTitle("Merge PDFs")
        self.setMinimumWidth(520)
        self._current_path = current_path
        self._page_count = page_count

        layout = QVBoxLayout(self)

        # File list
        layout.addWidget(QLabel("Files to merge (in order):"))
        self._file_list = QListWidget()
        self._file_list.setDragDropMode(QAbstractItemView.InternalMove)
        self._file_list.setDefaultDropAction(Qt.MoveAction)
        # Current doc is pinned at the top
        current_item = QListWidgetItem(f"[Current]  {current_path}")
        current_item.setFlags(current_item.flags() & ~Qt.ItemIsDragEnabled)
        self._file_list.addItem(current_item)
        layout.addWidget(self._file_list)

        # Add / remove buttons
        btn_row = QHBoxLayout()
        add_btn = QPushButton("+ Add PDF…")
        add_btn.clicked.connect(self._add_files)
        remove_btn = QPushButton("Remove selected")
        remove_btn.clicked.connect(self._remove_selected)
        btn_row.addWidget(add_btn)
        btn_row.addWidget(remove_btn)
        btn_row.addStretch()
        layout.addLayout(btn_row)

        # Insert position
        pos_group = QGroupBox("Insert added pages")
        pos_layout = QHBoxLayout(pos_group)
        self._pos_combo = QComboBox()
        self._pos_combo.addItem("After current document",  page_count)
        self._pos_combo.addItem("Before current document", 0)
        pos_layout.addWidget(self._pos_combo)
        pos_layout.addStretch()
        layout.addWidget(pos_group)

        # OK / Cancel
        buttons = QDialogButtonBox(QDialogButtonBox.Ok | QDialogButtonBox.Cancel)
        buttons.accepted.connect(self.accept)
        buttons.rejected.connect(self.reject)
        layout.addWidget(buttons)

    def _add_files(self):
        paths, _ = QFileDialog.getOpenFileNames(
            self, "Select PDFs to add", "", "PDF Files (*.pdf)"
        )
        for p in paths:
            self._file_list.addItem(QListWidgetItem(p))

    def _remove_selected(self):
        for item in self._file_list.selectedItems():
            # Do not allow removing the current-doc item (row 0)
            if self._file_list.row(item) == 0:
                continue
            self._file_list.takeItem(self._file_list.row(item))

    @property
    def other_paths(self) -> list[str]:
        paths = []
        for row in range(1, self._file_list.count()):
            paths.append(self._file_list.item(row).text())
        return paths

    @property
    def insert_at(self) -> int:
        return self._pos_combo.currentData()


# ── Split dialog ──────────────────────────────────────────────────────────────

class SplitDialog(QDialog):
    """Collect split parameters: mode, ranges/N, output folder, prefix."""

    def __init__(self, total_pages: int, parent=None):
        super().__init__(parent)
        self.setWindowTitle("Split PDF")
        self.setMinimumWidth(460)
        self._total = total_pages

        layout = QVBoxLayout(self)
        layout.addWidget(QLabel(f"Total pages: <b>{total_pages}</b>", textFormat=Qt.RichText))

        # Mode
        mode_group = QGroupBox("Split mode")
        mode_layout = QVBoxLayout(mode_group)
        self._mode_group = QButtonGroup(self)

        self._rb_ranges = QRadioButton("By page ranges  (e.g. 1-3, 4-7, 8-)")
        self._rb_every_n = QRadioButton("Every N pages")
        self._rb_each = QRadioButton("Each page as a separate file")
        self._rb_ranges.setChecked(True)

        for i, rb in enumerate([self._rb_ranges, self._rb_every_n, self._rb_each]):
            self._mode_group.addButton(rb, i)
            mode_layout.addWidget(rb)

        layout.addWidget(mode_group)

        # Ranges input
        form = QFormLayout()
        self._ranges_edit = QLineEdit()
        self._ranges_edit.setPlaceholderText(f"e.g. 1-3, 4-{total_pages}")
        form.addRow("Page ranges:", self._ranges_edit)

        self._n_spin = QSpinBox()
        self._n_spin.setMinimum(1)
        self._n_spin.setMaximum(total_pages)
        self._n_spin.setValue(1)
        form.addRow("Pages per file (N):", self._n_spin)

        layout.addLayout(form)

        # Output folder
        out_row = QHBoxLayout()
        self._out_edit = QLineEdit()
        self._out_edit.setPlaceholderText("Output folder…")
        browse_btn = QPushButton("Browse…")
        browse_btn.clicked.connect(self._browse_output)
        out_row.addWidget(self._out_edit)
        out_row.addWidget(browse_btn)
        layout.addLayout(out_row)

        # File name prefix
        self._prefix_edit = QLineEdit("output")
        prefix_row = QFormLayout()
        prefix_row.addRow("File name prefix:", self._prefix_edit)
        layout.addLayout(prefix_row)

        # OK / Cancel
        buttons = QDialogButtonBox(QDialogButtonBox.Ok | QDialogButtonBox.Cancel)
        buttons.accepted.connect(self._on_accept)
        buttons.rejected.connect(self.reject)
        layout.addWidget(buttons)

    def _browse_output(self):
        folder = QFileDialog.getExistingDirectory(self, "Choose output folder")
        if folder:
            self._out_edit.setText(folder)

    def _on_accept(self):
        if not self._out_edit.text().strip():
            QMessageBox.warning(self, "No folder", "Please choose an output folder.")
            return
        self.accept()

    @property
    def mode(self) -> str:
        return ["ranges", "every_n", "each"][self._mode_group.checkedId()]

    @property
    def ranges_text(self) -> str:
        return self._ranges_edit.text()

    @property
    def every_n(self) -> int:
        return self._n_spin.value()

    @property
    def out_dir(self) -> str:
        return self._out_edit.text().strip()

    @property
    def base_name(self) -> str:
        return self._prefix_edit.text().strip() or "output"


# ── Crop dialog ───────────────────────────────────────────────────────────────

class CropDialog(QDialog):
    """Collect crop margins in PDF points (72 pt = 1 inch)."""

    def __init__(self, page_count: int, parent=None):
        super().__init__(parent)
        self.setWindowTitle("Crop Pages")
        self.setMinimumWidth(360)

        layout = QVBoxLayout(self)
        layout.addWidget(
            QLabel("Margins to remove (PDF points — 72 pt = 1 inch):"),
        )

        form = QFormLayout()

        def _spin() -> QDoubleSpinBox:
            s = QDoubleSpinBox()
            s.setRange(0, 500)
            s.setDecimals(1)
            s.setSingleStep(5)
            s.setSuffix(" pt")
            return s

        self._top    = _spin()
        self._bottom = _spin()
        self._left   = _spin()
        self._right  = _spin()

        form.addRow("Top:",    self._top)
        form.addRow("Bottom:", self._bottom)
        form.addRow("Left:",   self._left)
        form.addRow("Right:",  self._right)
        layout.addLayout(form)

        # Apply to
        apply_group = QGroupBox("Apply to")
        apply_layout = QVBoxLayout(apply_group)
        self._apply_group = QButtonGroup(self)
        self._rb_current = QRadioButton("Current page only")
        self._rb_all     = QRadioButton("All pages")
        self._rb_range   = QRadioButton("Page range:")
        self._rb_current.setChecked(True)

        self._range_edit = QLineEdit()
        self._range_edit.setPlaceholderText(f"e.g. 1-{page_count}")
        self._range_edit.setEnabled(False)
        self._rb_range.toggled.connect(self._range_edit.setEnabled)

        for i, rb in enumerate([self._rb_current, self._rb_all, self._rb_range]):
            self._apply_group.addButton(rb, i)
            apply_layout.addWidget(rb)
        apply_layout.addWidget(self._range_edit)
        layout.addWidget(apply_group)

        layout.addWidget(QLabel(
            "<small>Visual crop overlay will be added in a future update.</small>",
            textFormat=Qt.RichText,
        ))

        buttons = QDialogButtonBox(QDialogButtonBox.Ok | QDialogButtonBox.Cancel)
        buttons.accepted.connect(self.accept)
        buttons.rejected.connect(self.reject)
        layout.addWidget(buttons)

    @property
    def margin_top(self)    -> float: return self._top.value()
    @property
    def margin_bottom(self) -> float: return self._bottom.value()
    @property
    def margin_left(self)   -> float: return self._left.value()
    @property
    def margin_right(self)  -> float: return self._right.value()

    @property
    def apply_to(self) -> str:
        return ["current", "all", "range"][self._apply_group.checkedId()]

    @property
    def range_text(self) -> str:
        return self._range_edit.text()
